<?xml version="1.0"?>
<launch>

  <arg name="robot_id" default="1" />
  <arg name="tf_prefix" value="robot$(arg robot_id)"/>

  <rosparam command="load" file="$(find orunav_params)/config/global.cfg"/>
  <rosparam command="load" file="$(find orunav_params)/config/citi1_simulation.cfg" />

  <include file="$(find cititruck_gazebo)/launch/cititruck_pallet_basement.launch"/>
  <include file="$(find gazebo_worlds_oru)/launch/basement_map_server.launch"/>

  <node pkg="orunav_vehicle_execution" type="point_n_click_target_client" name="point_n_click_target_client_citi1" output="screen">
    <param name="load_operation" value="true" />
    <param name="load_detect" value="true" />
    <rosparam param="robot_ids">[1]</rosparam>
    <param name="multiple_vehicles" value="true" />
  </node>

  <node pkg="orunav_coordinator_fake" type="coordinator_fake_node" name="coordinator_fake_node" output="screen">
    <param name="multiple_vehicles" value="true" />
    <param name="use_ct" value="false" />  
  </node>

  <!-- ALL ROBOT SPECIFIC TASKS -->

  <group ns="robot1">
  <param name="tf_prefix" value="robot1" />

  <node pkg="orunav_mpc" type="controller_node" name="controller" output="screen">
        <param name="robot_id" value="$(arg robot_id)" />
	<remap from="/$(arg tf_prefix)/cmd_vel" to ="/$(arg tf_prefix)/controller/cmd_vel"/>
  </node>
  <include file="$(find orunav_fork_control)/launch/fork_control_sim.launch"/>

  <node pkg="orunav_motion_planner" type="get_path_service" name="get_path_service" output="screen">		
    <param name="model" value="CiTiTruck_Expanded10+20_16_1_4.0_0.2.reduced"/>
    <param name="motion_primitives_directory" value="$(find orunav_motion_planner)/Primitives/" />
    <param name="lookup_tables_directory" value="$(find orunav_motion_planner)/LookupTables/" /> 
    <param name="min_incr_path_dist" value="0.2" />
    <param name="visualize" value="true" />
  </node>

  <node pkg="orunav_constraint_extract" type="polygonconstraint_service" name="polygonconstraint_service" output="screen">
    <param name="load_type" value="1" />
    <param name="model_type" value="2" />
    <param name="visualize" value="false" />
    <param name="save_constraints_and_path" value="true" />
    <param name="lookuptables_file" value="$(find orunav_constraint_extract)/tables/service_lookup_model2load1.dat" />
    <param name="visualize_outer_constraints" value="false" />
    <param name="visualize_only_invalid" value="false" />
    <param name="skip_overlap" value="true" />
    <param name="save_lookuptables" value="false" />
  </node>  

  <node pkg="orunav_path_smoother" type="smoothed_path_service" name="smoothed_path_service" output="screen">
    <param name="visualize" value="true" />
    <param name="visualize_deep" value="false" />
    <param name="wheel_base" value="1.190" />
    <param name="minimize_phi_and_dist" value="true" />
    <param name="init_controls" value="true" />
    <param name="use_th_constraints" value="true" />
    <param name="use_xy_constraints" value="true" />
    <param name="phi_min" value="-1.0" /> <!-- -0.9 -->
    <param name="phi_max" value="1.0" />  <!-- 0.9 -->
    <param name="reassign_constraints" value="true" />
    <param name="reassign_iters" value="1" />
    <param name="reassign_min_distance" value="0.1" />
  </node>  

  <node pkg="orunav_vehicle_execution" type="orunav_vehicle_execution_node" name="orunav_vehicle_execution_node_citi1" output="screen">
    <param name="visualize" value="true" />
    <param name="max_tracking_error" value="-1." />
    <param name="use_forks" value="true" />

    <param name="max_vel" value="0.3" />
    <param name="max_rotational_vel" value="0.3" />
    <param name="max_acc" value="0.1" />
  </node>

    <node pkg="orunav_pallet_detection_sdf" type="euro_pallet_sdf_node" name="euro_pallet_sdf_node">
      <param name="background_indices_file_name" value="$(find orunav_pallet_detection_sdf)/backgrounds/citi1_sim_background.dat"/>
      <param name="Dmax" value="0.2"/>
      <param name="visualize" value="true"/>
      <param name="visualize_sdf" value="true"/>
      <param name="removeFloor" value="true"/>
      <param name="sensor_time_offset" value="0.0"/>
      <remap from="pointcloud" to ="kinect_rgb/sensors/kinect2depth_registered/points"/>
      <param name="residual_thresh" value="0.01"/>
      <param name="cov_thresh" value="0.2"/>
      <param name="only_compute_filter_used" value="true"/>
      <param name="depth_frame_id" value="robot1/kinect_link" /> <!-- robot1/kinect_link -->
      <param name="asus_camera" value="false" />
      <param name="min_nb_matched_points" value="10000" />
      <param name="min_match_nb_points_ratio" value="0.1" />
    </node>

    <node pkg="orunav_path_smoother" type="smoothed_straight_path_service" name="smoothed_straight_path_service" output="screen">
      <param name="visualize" value="true" />
    </node>  

    <node pkg="orunav_vehicle_execution" type="vehicle_execution_client" name="citi1_vehicle_client" output="screen">
      <param name="targets_file_name" value="$(find orunav_vehicle_execution)/targets/basement_pallet_targets.dat" />
    </node>

    
  </group>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find cititruck_description)/launch/click_n_pick.rviz"/>

</launch>
